src/data_device_manager/data_device.rs

```diff
@@ -0,0 +1,218 @@
+pub struct DataDeviceData {
+    pub(super) inner: Mutex<DataDeviceInner>,
+}
+
+pub trait DataDeviceDataExt: Send + Sync {
+    fn data_device_data(&self) -> &DataDeviceData;
+
+    /// get the active dnd offer if it exists
+    fn dnd_offer(&self) -> Option<DndDataOffer> {
+        self.data_device_data().inner.lock().ok().and_then(|inner| inner.dnd_data_offer.clone())
+    }
+
+    /// get the active selection offer if it exists
+    fn selection_offer(&self) -> Option<WlDataOffer> {
+        self.data_device_data().inner.lock().ok().and_then(|inner| inner.selection.clone())
+    }
+}
+
+impl DataDeviceDataExt for DataDeviceData {
+    fn data_device_data(&self) -> &DataDeviceData {
+        self
+    }
+}
+
+/// Handler trait for DataDevice events.
+///
+/// The functions defined in this trait are called as DataDevice events are received from the compositor.
+pub trait DataDeviceHandler: Sized {
+    /// Introduces a new data offer
+    fn data_offer(
+        &mut self,
+        conn: &Connection,
+        qh: &QueueHandle<Self>,
+        data_device: &WlDataDevice,
+        offer: WlDataOffer,
+    );
+
+    /// The data device pointer has entered a surface at the provided location
+    #[allow(clippy::too_many_arguments)]
+    fn enter(&mut self, conn: &Connection, qh: &QueueHandle<Self>, data_device: &WlDataDevice);
+
+    /// The drag and drop pointer has left the surface and the session ends
+    /// The offer will be destroyed
+    fn leave(&mut self, conn: &Connection, qh: &QueueHandle<Self>, data_device: &WlDataDevice);
+
+    /// Drag and Drop motion
+    fn motion(&mut self, conn: &Connection, qh: &QueueHandle<Self>, data_device: &WlDataDevice);
+
+    /// Advertises a new selection
+    fn selection(&mut self, conn: &Connection, qh: &QueueHandle<Self>, data_device: &WlDataDevice);
+
+    /// drop performed
+    /// after the next data offer action event, data may be able to be received, unless the action is "ask"
+    fn drop_performed(
+        &mut self,
+        conn: &Connection,
+        qh: &QueueHandle<Self>,
+        data_device: &WlDataDevice,
+    );
+}
+
+impl<D, U> Dispatch<wl_data_device::WlDataDevice, U, D> for DataDeviceManagerState
+where
+    D: Dispatch<wl_data_device::WlDataDevice, U>
+        + Dispatch<wl_data_offer::WlDataOffer, DataOfferData>
+        + DataDeviceHandler
+        + DataOfferHandler
+        + 'static,
+    U: DataDeviceDataExt,
+{
+    event_created_child!(D, WlDataDevice, [
+        0 => (WlDataOffer, DataOfferData::default())
+    ]);
+
+    fn event(
+        state: &mut D,
+        data_device: &wl_data_device::WlDataDevice,
+        event: wl_data_device::Event,
+        data: &U,
+        conn: &Connection,
+        qh: &QueueHandle<D>,
+    ) {
+        let data = data.data_device_data();
+        let mut inner = data.inner.lock().unwrap();
+
+        match event {
+            wayland_client::protocol::wl_data_device::Event::DataOffer { id } => {
+                drop(inner);
```

Comment:
We should note this drop is required where it's used in this dispatch impl to prevent Mutex deadlocks.